[expo-app-loading][expo] Extract AppLoading to separate module

bbarthec opened the pull request:
# Why

`AppLoading` is the last module in Expo that can be extracted into the separate module. Although `expo-splash-screen` is now our source of truth about splash-related things, to keep backward compatibility we need this package as well.
Additionally, we want to remove `expo-splash-screen` dependency from the `expo` module, because it's being auto-installed in bare workflow when the user is using `expo` package and we don't want that (it has unwanted side-effects that are sometimes unwanted).

# How

- created a separate `expo-app-loading` module in `packages` and moved `AppLoading` code there. This packages is a single-react-component package now.
- removed `RootErrorBoundary` as it's now no-op component on every platform (previously it was only affecting the Android platform)
- tweaked & sorted tests and `removed/deprecated` packages lists
- corrected all imports of `{ AppLoading }` or replaced it with `SplashScreen` if possible

# Test Plan

- `expo` tests are passing
- managed `ncl` & `bare-expo` apps are working
- when using the `AppLoading` according to the current example ([example](https://docs.expo.io/versions/latest/sdk/app-loading/#props) `import { AppLoading } from 'expo'`) the developer will encounter the well-known error message in the console:
```
Error: The following APIs have moved to separate packages and importing them from the "expo" package is no longer supported: AppLoading.

1. Add correct versions of these packages to your project using:

   expo install expo-app-loading

   If "install" is not recognized as an expo command, update your expo-cli installation.

2. Change your imports so they use specific packages instead of the "expo" package:

 - import { AppLoading } from 'expo' -> import AppLoading from 'expo-app-loading'
```
- after changing the import according to the suggestion everything works as previously 🎉

bbarthec committed:
[expo-app-loading] Create module by extracting it from expo module

bbarthec committed:
[expo] Remove AppLoading component from expo module

bbarthec committed:
[babel-preset-expo] Replace expo.AppLoading with AppLoading

bbarthec committed:
[expo] Remove RootErrorBoundary

I've tracked down that this component is no-op:
- [x] it is not a part of public API
- [x] it was used only on Android
- [x] all it was doing was to hide the splash screen upon error and this is no longer an issue on Android (in fact the native module responsible for this job is no longer exposed from the Android native realm 😅)

bbarthec committed:
[expo] Remove AppLoading/RootErrorBoundary tests

bbarthec committed:
[ncl] Replace AppLoading with SplashScreen

bbarthec committed:
[CHANGELOG] Add CHANGELOG entry

bbarthec committed:
[CODEOWNERS] Add @bbarthec as a expo-app-loading CODEOWNER

lukmccall reviewed (approved):
LGTM 🚀🚀🚀

ide reviewed (changes requested):
The autoHideAsync prop needs rethinking and might be best for a separate PR since this PR is already doing a lot.

I would also like to see a more detailed test plan that explains how to test the various code paths and edge cases and what behavior to look for.

brentvatne commented on packages/expo-app-loading/src/AppLoading.tsx:
```diff
@@ -0,0 +1,101 @@
+  /**
+   * Whether to hide the native splash screen as soon as you unmount the `AppLoading` component.
+   */
+  autoHideAsync?: boolean;
```
yeah this looks like a mistake maybe?

brentvatne replied:
it should be `autoHideSplash` - it's used here: https://github.com/expo/expo/blob/6ebc84226e56e7672c9e08137f4c2a1e0bb85891/packages/expo-app-loading/src/AppLoadingNativeWrapper.tsx#L19

brentvatne commented on packages/expo-app-loading/src/AppLoading.tsx:
```diff
@@ -0,0 +1,101 @@
+    emitEvent('componentWillUnmount');
+  }
+
+  private startLoadingAppResourcesAsync = async () => {
```
`private async startLoadingAppResourcesAsync() {` 😅

brentvatne commented on packages/expo/src/removed.ts:
```diff
@@ -698,6 +710,17 @@ Object.defineProperties(module.exports, {
     },
   },
   /** @deprecated */
+  Pedometer: {
```
usually best to keep changes like this to separate commits, same for reordering lines from removed.web.ts below.

brentvatne commented on packages/expo-app-loading/src/AppLoading.tsx:
```diff
@@ -0,0 +1,101 @@
+   * This should be used to set state and unmount the `AppLoading` component.
+   * @deprecated
+   */
+  onFinish?: () => void;
```
`startAsync`, `onError`, and `onFinish` are all meant to be optional because you could just mount the `AppLoading` component without any props (this is how we used to recommend using it) and do your own work in `useEffect` or something.

bbarthec committed:
[expo-app-loading] Apply code suggestions

bbarthec commented on packages/expo/src/__tests__/Expo-test.native.ts:
```diff
@@ -2,8 +2,6 @@
 // when we add or change a native module and haven't yet updated the mocks in jest-expo. This list
 // is a temporary workaround, not a way to indefinitely avoid testing modules.
 const skippedExports: string[] = [
-  'IntentLauncherAndroid',
-  'Localization',
```
These are duplicated below.

bbarthec committed:
[expo-app-loading] Replace isMounted with _isMounted

`this.isMounted` is somehow internally used by ReactNative and basing on such named property brings no good, only warnings.

bbarthec committed:
[expo-app-loading] Fix CI errors

brentvatne commented on packages/expo-app-loading/README.md:
```diff
@@ -0,0 +1,27 @@
+# expo-app-loading
+
+A React component that tells Expo to keep the app loading screen open if it is the first and only component rendered in your app. Unless the `autoHideSplash` prop is set to `false`, the loading screen will disappear and your app will be visible when the component is removed.
```
```suggestion
A React component that tells expo-splash-screen to remain visible if it is the first and only component rendered in your app.
```

brentvatne commented on packages/expo-app-loading/README.md:
```diff
@@ -0,0 +1,27 @@
+# expo-app-loading
+
+A React component that tells Expo to keep the app loading screen open if it is the first and only component rendered in your app. Unless the `autoHideSplash` prop is set to `false`, the loading screen will disappear and your app will be visible when the component is removed.
+This is incredibly useful to let you download and cache fonts, logos, icon images and other assets that you want to be sure the user has on their device for an optimal experience before rendering and they start using the app.
```
```suggestion
This can be useful while download and cache fonts, logos, icon images and other assets that you want to be sure the user has on their device for an optimal experience before rendering and they start using the app. You can alternatively use expo-splash-screen APIs directly - expo-app-loading just wraps them.
```

brentvatne commented on packages/expo-app-loading/src/AppLoading.tsx:
```diff
@@ -0,0 +1,108 @@
+  }
+}
+
+export function getAppLoadingLifecycleEmitter(): EventEmitter {
```
i think we can remove all of this event emitter stuff since we are removing the error boundary component

brentvatne commented on packages/expo-app-loading/src/AppLoadingNativeWrapper.tsx:
```diff
@@ -6,21 +6,25 @@ type Props = {
 
 export default class AppLoading extends React.Component<Props> {
+  static defaultProps = {
+    autoHideAsync: true,
```
```suggestion
    autoHideSplash: true,
```

brentvatne committed:
Apply suggestions from code review

brentvatne commented:
@bbarthec - can you open a follow up PR that removes the app lifecycle event emitter? :)

bbarthec committed:
[ci] Fix packages

bbarthec committed:
[ci][babel-preset-expo] Remove unsupported decorators

bbarthec committed:
[ci][babel-preset-expo][jest] Update snapshots

egecavusoglu commented:
I was upgrading to Expo SDK 40, and followed the changes that need to be done for this module. (simply changed the import)
import AppLoading from "expo-app-loading";
```
AppLoading threw an unexpected error when loading:
startLoadingAppResourcesAsync$@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:134432:32
tryCatch@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:28978:23
invoke@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:29151:32
tryCatch@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:28978:23
invoke@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:29051:30
http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:29081:19
tryCallTwo@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:3879:9
doResolve@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:4043:25
Promise@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:3902:14
callInvokeWithMethodAndArg@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:29080:33
enqueue@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:29085:157
http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:29102:69
startLoadingAppResourcesAsync@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:134407:42
componentDidMount@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:134395:43
commitLifeCycles@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:23707:47
commitLayoutEffects@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:25967:29
invokeGuardedCallbackImpl@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:13188:21
invokeGuardedCallback@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:13282:42
commitRootImpl@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:25756:36
commitRootImpl@[native code]
unstable_runWithPriority@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:47595:30
commitRoot@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:25643:24
performSyncWorkOnRoot@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:25167:19
performSyncWorkOnRoot@[native code]
http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:16439:38
unstable_runWithPriority@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:47595:30
flushSyncCallbackQueueImpl@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:16434:28
flushSyncCallbackQueue@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:16423:35
scheduleUpdateOnFiber@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:24776:37
updateContainer@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:27255:21
render@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:27689:24
renderApplication@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:75066:77
run@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:74636:28
runApplication@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:74693:28
__callFunction@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:3193:36
http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:2925:31
__guard@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:3147:15
callFunctionReturnFlushedQueue@http://127.0.0.1:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&minify=false:2924:21
callFunctionReturnFlushedQueue@[native code]
- node_modules/react-native/Libraries/LogBox/LogBox.js:148:8 in registerError
- node_modules/react-native/Libraries/LogBox/LogBox.js:59:8 in errorImpl
- node_modules/react-native/Libraries/LogBox/LogBox.js:33:4 in console.error
- node_modules/expo/build/environment/react-native-logs.fx.js:27:4 in error
- node_modules/expo-app-loading/build/AppLoading.js:10:13 in componentDidMount
- node_modules/react-native/node_modules/promise/setimmediate/core.js:37:13 in tryCallOne
- node_modules/react-native/node_modules/promise/setimmediate/core.js:123:24 in setImmediate$argument_0
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:130:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:181:14 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:441:30 in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:387:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:135:6 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:364:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:134:4 in flushedQueue
* [native code]:null in flushedQueue
* [native code]:null in callFunctionReturnFlushedQueue
```

Any insights on this?

egecavusoglu commented:
Nvm, in debug mode found out this message which is helpful enough.

```
Error: AppLoading onError prop is required if startAsync is provided
```
